rust/src/dataset/optimize.rs

```diff
@@ -0,0 +1,1015 @@
+        } else {
+            0.0
+        }
+    }
+
+    /// The number of rows that are still in the fragment
+    fn num_rows(&self) -> usize {
+        self.fragment_length - self.num_deletions
+    }
+}
+
+async fn collect_metrics(fragment: &FileFragment) -> Result<FragmentMetrics> {
+    let fragment_length = fragment.fragment_length();
+    let num_deletions = fragment.count_deletions();
+    let (fragment_length, num_deletions) =
+        futures::future::try_join(fragment_length, num_deletions).await?;
+    Ok(FragmentMetrics {
+        fragment_length,
+        num_deletions,
+    })
+}
+
+/// A plan for what groups of fragments to compact.
+///
+/// See [plan_compaction()] for more details.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CompactionPlan {
+    tasks: Vec<TaskData>,
+    read_version: u64,
+    options: CompactionOptions,
+}
+
+impl CompactionPlan {
+    /// Retrieve standalone tasks that be be executed in a distributed fashion.
+    pub fn compaction_tasks(&self) -> impl Iterator<Item = CompactionTask> + '_ {
+        let read_version = self.read_version;
+        let options = self.options.clone();
+        self.tasks.iter().map(move |task| CompactionTask {
+            task: task.clone(),
+            read_version,
+            options: options.clone(),
+        })
+    }
+
+    /// The number of tasks in the plan.
+    pub fn num_tasks(&self) -> usize {
+        self.tasks.len()
+    }
+
+    /// The version of the dataset that was read to produce this plan.
+    pub fn read_version(&self) -> u64 {
+        self.read_version
+    }
+
+    /// The options used to produce this plan.
+    pub fn options(&self) -> &CompactionOptions {
+        &self.options
+    }
+}
+
+/// A single group of fragments to compact, which is a view into the compaction
+/// plan. We keep the `replace_range` indices so we can map the result of the
+/// compact back to the fragments it replaces.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct TaskData {
+    /// The fragments to compact.
+    pub fragments: Vec<Fragment>,
+}
+
+/// A standalone task that can be serialized and sent to another machine for
+/// execution.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct CompactionTask {
+    pub task: TaskData,
+    pub read_version: u64,
+    options: CompactionOptions,
+}
+
+impl CompactionTask {
+    /// Run the compaction task and return the result.
+    ///
+    /// This result should be later passed to [commit_compaction()] to commit
+    /// the changes to the dataset.
+    pub async fn execute(&self, dataset: &Dataset) -> Result<RewriteResult> {
+        if dataset.manifest.version != self.read_version {
+            return Err(crate::Error::invalid_input(format!(
+                "Compaction was planned for version {}, but dataset is now at version {}.",
+                self.read_version, dataset.manifest.version
+            )));
+        }
```

Comment:
This is probably something worth mentioning in the docs somewhere.  Even the slightest update to a dataset will cause all compaction tasks to fail.  Users should probably be mindful and run the compaction during a quiet period of some kind.

Reply:
Hmm I think I could actually loosen this maybe. We should be compatible with Append and non-overlapping deletes file:///Users/willjones/Documents/lance/rust/target/doc/lance/dataset/transaction/index.html

Reply:
I've changed it so it switches the the correct version if needed.